### PR TITLE
feat: implementing alerts for typed sign data signatures

### LINKED ISSUE #24611: Implement alerts system for typed signatures
Following Vini's work for personal signature, we need to extend alert system to typed signatures also.

Ref: https://github.com/MetaMask/MetaMask-planning/issues/2412

### PR DESCRIPTION
**Pull request by jpuri**
## **Description**

Extend alert system to typed signatures.

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/24611

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/pages/confirmations/utils/confirm.ts
Comment (matthewwalsh0) [inline ui/pages/confirmations/utils/confirm.ts:28]: In the case of the signature confirmations, is the `type` property actually an `ApprovalType` in which case would it be clearer to define this list using the `ApprovalType` values to decouple it from transactions?

**Review by pedronfigueiredo — DISMISSED** [review]

File: ui/pages/confirmations/hooks/setConfirmationAlerts.ts
Comment (jpuri) [reply ui/pages/confirmations/hooks/setConfirmationAlerts.ts:14]: The only thing `setConfirmationAlerts` has more than `useConfirmationAlerts` is updating the state. Thus 2 hooks here look bit redundant.

File: ui/pages/confirmations/utils/confirm.ts
Comment (jpuri) [reply ui/pages/confirmations/utils/confirm.ts:28]: I added this method to be used for `currentConfirmation` object, it is not approval request. It comes from MessageManagers / TransactionController - thus I was not sure if its ok to use ApprovalType.

File: ui/components/app/alerts-system/utils.ts
Comment (jpuri) [reply ui/components/app/alerts-system/utils.ts:103]: Yeh no benefit actually, only a preference. I can rename and revert this change.

File: ui/pages/confirmations/hooks/setConfirmationAlerts.ts
Comment (jpuri) [reply ui/pages/confirmations/hooks/setConfirmationAlerts.ts:14]: For current confirmation we have 2 hooks `useCurrentConfirmation` and `setCurrentConfirmation` as `useCurrentConfirmation` has a use case outside `setCurrentConfirmation`.

Comment (matthewwalsh0) [reply ui/pages/confirmations/hooks/setConfirmationAlerts.ts:14]: Updating the state is the central purpose of `setConfirmationAlerts` though.

My concern is that in future we could have many many hooks for each approval type for example, and we'd have to reference each one directly within this hook, but it would be cleaner and easier to maintain them in an aggregate hook like `useConfirmationAlerts`.

File: ui/pages/confirmations/utils/confirm.ts
Comment (matthewwalsh0) [reply ui/pages/confirmations/utils/confirm.ts:28]: I see what you mean.

This makes sense for now but I think it highlights a need to move away from a confirmation directly being alternate controller state objects, and instead define a standardised confirmation type that uses an `ApprovalType` for type and a generic `data` field for anything specific to the type.

This would be very similar to the `ApprovalRequest` so we could even just re-use that.

File: ui/pages/confirmations/hooks/setConfirmationAlerts.ts
Comment (jpuri) [reply ui/pages/confirmations/hooks/setConfirmationAlerts.ts:14]: I see, let me revert this change.

File: ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx
Comment (vinistevam) [inline ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx:124]: :heart:

File: ui/pages/confirmations/components/confirm/footer/footer.tsx
Comment (vinistevam) [inline ui/pages/confirmations/components/confirm/footer/footer.tsx:78]: After our discussion regarding the designs for [the MVP](https://www.figma.com/design/yUczlH2HbHaHsYijiONKx0/Warnings%2C-Alerts-and-Flags?node-id=2113%3A40681&t=Ac1b3BuZOUUrsP54-1) , I wanted to confirm a few details.
Once we receive a general alert from Blockaid, the button will be labeled `Review Alerts`. Clicking this button will open a friction modal where the user can acknowledge the alert and proceed to confirm the action.

Is that correct?

File: ui/components/app/alert-system/alert-modal/alert-modal.test.tsx
Comment (vinistevam) [inline ui/components/app/alert-system/alert-modal/alert-modal.test.tsx:169]: nit: Just to remove the `any` we could use `(useAlertsSpy as jest.Mock).mockReturnValue({`

File: ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.test.tsx
Comment (vinistevam) [inline ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.test.tsx:96]: Do we need to comment out these two unit tests? I understand that the MVP does not support multiple providers, but can we keep the use case in the unit tests unless they are broken due to the MVP changes?

File: ui/pages/confirmations/components/confirm/footer/footer.tsx
Comment (jpuri) [reply ui/pages/confirmations/components/confirm/footer/footer.tsx:78]: Yes, that is how it will work.

File: ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.test.tsx
Comment (jpuri) [reply ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.test.tsx:96]: Yep that is issue these test are breaking.

File: ui/components/app/alert-system/alert-modal/alert-modal.tsx
Comment (matthewwalsh0) [inline ui/components/app/alert-system/alert-modal/alert-modal.tsx:318]: Why does this need to be hardcoded inside the `AlertModal` which aims to be decoupled from any single alert or domain?

Could the `blockaidAlertInfo` label instead be used when the alert is generated from the relevant `useConfirmationAlerts` hook so it's not a special case, but a standard alert like all others?

File: ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx
Comment (matthewwalsh0) [inline ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx:178]: As mentioned before, we want the alert system to be generic and decoupled from domain. If we add clauses like this, we're increasing that coupling and the level of complexity given the additional unique behaviours.

If we need custom text here for some alerts, could we keep it generic and add a suitable optional property to the `Alert` type itself?

Comment (matthewwalsh0) [inline ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx:178]: As an aside, should this ternary be comparing against the `selectedAlert.provider`?

This appears to be checking if an enum property exists which will always be `true`?

File: ui/pages/confirmations/components/confirm/footer/footer.tsx
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/footer/footer.tsx:26]: As discussed before, I'd vote to use the `function` syntax for traditional top level functions.

For the components themselves, I appreciate it's personal preference whether we go `const` or `function`.

**Comment by bschorchit** [comment]
Testing the latest build on this PR, I see malicious warning showing up as blue with a blue Confirm button first before both the banner and button change to red.
Screenshot 2024-05-29 at 08 49 56

File: ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx
Comment (jpuri) [reply ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx:178]: I totally agree on this point. My understanding was that how blockaid alerts work currently is temporary solution we arrived on to un-block alerts for signatures. Eventually it should be done in generic way.

If that is not the case definitely we should do it in generic way.

File: ui/components/app/alert-system/alert-modal/alert-modal.tsx
Comment (jpuri) [reply ui/components/app/alert-system/alert-modal/alert-modal.tsx:318]: Please check this: https://github.com/MetaMask/metamask-extension/pull/24702#discussion_r1618805115

File: ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx
Comment (jpuri) [reply ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx]: Fixed 👍

**Comment by jpuri** [comment]
Hey @bschorchit : I tested it using malicious permits in test dapp and it seems to work well:

Screenshot 2024-05-29 at 6 39 13 PM

**Comment by SayaGT** [comment]
@bschorchit @jpuri yes the font size for the alert details seems wrong. It should be Body-MD (baseline).

**Comment by bschorchit** [comment]
> Hey @bschorchit : I tested it using malicious permits in test dapp and it seems to work well:

I do see it as red as well, but sometimes only after ~1s. It doesn't happen every time, but I can consistently reproduce by trying a few times and alternating between `Malicious Permit`, `Malicious Trade Order`, `Malicious Seaport`

Here's a recording:

https://github.com/MetaMask/metamask-extension/assets/34968356/d1939d0f-9c99-4399-a985-fce91e7da2ac

**Comment by bschorchit** [comment]
Summing up to facilitate reading (sorry for the multiple comments):
1. Font size for alert details should be Body-MD (baseline) - [source](https://github.com/MetaMask/metamask-extension/pull/24702#issuecomment-2137463113)
2. CTA should be `Confirm` rather than `Review alerts` as this is the single alert scenario - [source](https://github.com/MetaMask/metamask-extension/pull/24702#issuecomment-2137463113)
3. Potential issue with rendering blue banner first - [source](https://github.com/MetaMask/metamask-extension/pull/24702#issuecomment-2137810388)
4. Small copy updates to the friction modal - (sorry this is new, and totally on me for not getting content frozen sooner. I'll get the final copy soon) - we can tackle this on a follow up PR to prevent blocking this one from being blocked.

File: ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx
Comment (matthewwalsh0) [reply ui/components/app/alert-system/confirm-alert-modal/confirm-alert-modal.tsx]: If it's only temporary, then that makes total sense but should we create a ticket to log the future refactor to make it more generic?

**Review by matthewwalsh0 — APPROVED** [review]

**Review by vinistevam — APPROVED** [review]
Amazing PR! :1st_place_medal: :rocket:

**Pull request merged**
